editor/src/core/model/element-metadata-utils.ts

```diff
@@ -1362,6 +1370,78 @@ export const MetadataUtils = {
   },
 }
 
+function fillSpyOnlyMetadataWithFramesFromChildren(
+  fromSpy: ElementInstanceMetadataMap,
+  fromDOM: ElementInstanceMetadataMap,
+) {
+  const childrenInDomCache: { [pathStr: string]: Array<ElementInstanceMetadata> } = {}
+
+  const findChildrenInDomRecursively = (pathStr: string): Array<ElementInstanceMetadata> => {
+    const existing = childrenInDomCache[pathStr]
+
+    if (existing != null) {
+      return existing
+    }
+
+    const spyElem = fromSpy[pathStr]
+    const childrenFromSpy = MetadataUtils.getChildren(fromSpy, spyElem.elementPath)
+    const childrenFromDom = MetadataUtils.getChildren(fromDOM, spyElem.elementPath)
+    const childrenNotInDom = childrenFromSpy.filter((childNotInDom) =>
+      childrenFromDom.every(
+        (childInDom) => !EP.pathsEqual(childNotInDom.elementPath, childInDom.elementPath),
+      ),
+    )
+    const recursiveChildren = childrenNotInDom.flatMap((c) => {
+      return findChildrenInDomRecursively(EP.toString(c.elementPath))
+    })
+    const children = [...childrenFromDom, ...recursiveChildren]
+
+    childrenInDomCache[pathStr] = children
+
+    return children
+  }
+
+  // those elements which are not in the dom need calculated localFrame and globalFrame
+  // from their children (or deeper descendants)
+  const elementsWithoutDomMetadata = Object.keys(fromSpy).filter((p) => fromDOM[p] == null)
+
+  const workingElements: ElementInstanceMetadataMap = {}
+
+  fastForEach(elementsWithoutDomMetadata, (pathStr) => {
+    const spyElem = fromSpy[pathStr]
+    const children = findChildrenInDomRecursively(pathStr)
+    if (children.length === 0) {
+      return
+    }
+    let workingMetadata: ElementInstanceMetadata = { ...spyElem }
+
+    fastForEach(children, (childMetadata) => {
+      const globalFrame =
+        workingMetadata.globalFrame != null
+          ? boundingRectangle(
+              workingMetadata.globalFrame ?? zeroCanvasRect,
```

Comment:
Annoying that TS didn't calculate that this `??` isn't necessary here

Reply:
Not relevant after f569e233e38e07734beded4ab85186b194c3fe8b